internal/js_parser/js_parser.go

```diff
@@ -7702,6 +7702,36 @@ func (p *parser) mangleStmts(stmts []js_ast.Stmt, kind stmtsKind) []js_ast.Stmt
 			}
 
 		case *js_ast.SLocal:
+			// Drop unused local variables
+			//
+			// The declaration must not be exported. We can't just check for the
+			// "export" keyword because something might do "export {id};" later on.
+			// Instead we just ignore all top-level declarations for now. That means
+			// this optimization currently only applies in nested scopes.
+			//
+			// Ignore declarations if the scope is shadowed by a direct "eval" call.
+			// The eval'd code may indirectly reference this symbol and the actual
+			// use count may be greater than 0.
+			if p.currentScope != p.moduleScope && !p.currentScope.ContainsDirectEval {
```

Comment:
If this costs too much, I think this condition could be `p.currentScope.Parent == p.moduleScope && !p.currentScope.ContainsDirectEval` instead.